stl/src/tzdb.cpp

```diff
@@ -1,15 +1,297 @@
+        // clang-format on
+    };
+
+    _NODISCARD const char* _Allocate_wide_to_narrow(
+        const char16_t* _Input, int _Input_len, __std_tzdb_error& _Err) noexcept {
+        const auto _Code_page      = __std_fs_code_page();
+        const auto _Input_as_wchar = reinterpret_cast<const wchar_t*>(_Input);
+        // FIXME: Is is ok to pull in xfilesystem_abi.h and use these here?
+        const auto _Count_result = __std_fs_convert_wide_to_narrow(_Code_page, _Input_as_wchar, _Input_len, nullptr, 0);
+        if (_Count_result._Err != __std_win_error::_Success) {
+            _Err = __std_tzdb_error::_Win_error;
+            return nullptr;
+        }
+
+        auto* _Data = new (_STD nothrow) char[_Count_result._Len + 1];
+        if (_Data == nullptr) {
+            return nullptr;
+        }
+
+        _Data[_Count_result._Len] = '\0';
+
+        const auto _Result =
+            __std_fs_convert_wide_to_narrow(_Code_page, _Input_as_wchar, _Input_len, _Data, _Count_result._Len);
+        if (_Result._Err != __std_win_error::_Success) {
+            _Err = __std_tzdb_error::_Win_error;
+            delete[] _Data;
+            return nullptr;
+        }
+
+        return _Data;
+    }
+
+} // namespace
+
 _EXTERN_C
 
+_NODISCARD __std_tzdb_time_zones_info* __stdcall __std_tzdb_get_time_zones() noexcept {
+    // On exit---
+    //    *_Info == nullptr         --> bad_alloc()
+    //    _Info->_Err == _Win_error --> failed, call GetLastError()
+    //    _Info->_Err == _Icu_error --> runtime_error interacting with ICU
+    _STD unique_ptr<__std_tzdb_time_zones_info, decltype(&__std_tzdb_delete_time_zones)> _Info{
+        new (_STD nothrow) __std_tzdb_time_zones_info{}, &__std_tzdb_delete_time_zones};
+    if (_Info == nullptr) {
+        return nullptr;
+    }
+
+    if (_Acquire_icu_functions() < _Icu_api_level::__has_icu_addresses) {
+        _Info->_Err = __std_tzdb_error::_Win_error;
+        return _Info.release();
+    }
+
+    UErrorCode _Err{};
+    _STD unique_ptr<UEnumeration, decltype(&__icu_uenum_close)> _Enum{
+        __icu_ucal_openTimeZoneIDEnumeration(USystemTimeZoneType::UCAL_ZONE_TYPE_CANONICAL, nullptr, nullptr, &_Err),
+        &__icu_uenum_close};
+    if (U_FAILURE(_Err)) {
+        _Info->_Err = __std_tzdb_error::_Icu_error;
+        return _Info.release();
+    }
+
+    // uenum_count may be expensive but is required to pre allocated arrays.
+    int32_t _Num_time_zones = __icu_uenum_count(_Enum.get(), &_Err);
+    if (U_FAILURE(_Err)) {
+        _Info->_Err = __std_tzdb_error::_Icu_error;
+        return _Info.release();
+    }
+
+    _Info->_Num_time_zones = static_cast<size_t>(_Num_time_zones);
+    _Info->_Names          = new (_STD nothrow) const char*[_Info->_Num_time_zones];
+    if (_Info->_Names == nullptr) {
+        return nullptr;
+    }
+
+    // init to ensure __std_tzdb_delete_init_info cleanup is valid
+    _STD fill_n(_Info->_Names, (ptrdiff_t) _Info->_Num_time_zones, nullptr);

```

Comment:
We conventionally avoid C-style casts, as they are difficult to search for and are willing to do surprising things.